### PR TITLE
Lesson 4 tasks

### PR DESCRIPTION
**Pull request by nazarenkod**

### CODE REVIEW

File: test/DomainRegistry.test.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
+const { expect } = require("chai");
+
+describe("DomainRegistry", function () {
+  let domainRegistry; // Объявляем переменную для контракта
+  let tldToRegister; // Объявляем переменную для домена
+  let owner; // Объявляем переменную для аккаунта владельца
```
Comment (AndriianChestnykh) [inline test/DomainRegistry.test.js]: Do you really need to comment what is obvious? If you work with ChatGPT that does not mean you should just copy-paste.
 -5

File: test/DomainRegistry.test.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    const DomainRegistry = await hre.ethers.getContractFactory("DomainRegistry");
+
+    // Деплоим контракт DomainRegistry и сохраняем его в переменной
+    domainRegistry = await DomainRegistry.deploy();
+
+    // Получаем аккаунты из Hardhat
+    [owner] = await hre.ethers.getSigners();
+
+    // Параметры для регистрации домена
+    tldToRegister = "com"; // Убедитесь, что tldToRegister установлен в корректное значение
```
Comment (AndriianChestnykh) [inline test/DomainRegistry.test.js]: Can `tldToRegister` have a better name? -1

File: contracts/DomainRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    
+
+    struct Domain {
+        address controller; 
+        string tld; 
+        uint256 createdTimestamp; 
+        uint256 deposit; 
+        bool isRegistered; 
+        
+    }
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol]: Can transactions be cheaper if optimise this struct? -3

File: contracts/DomainRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+
+    // Мапа для зберігання доменів, де ключ - це назва домену верхнього рівня
+    mapping(string => Domain) public domains;
+
+    // Подія, яка виникає при створенні нового домену
+    event DomainCreated(address indexed controller, string indexed tld, uint256 createdTimestamp, uint256 deposit);
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol]: Any excessive fields in this event? -2

File: contracts/DomainRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: MIT
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:1]: Does this contract covers all most probable questions in terms of data monitoring? -3

File: contracts/DomainRegistry.sol [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    modifier domainDoesNotExist(string memory _tld) {
+        require(!domains[_tld].isRegistered, "Domain exists");
+        _;
+    }
+
+modifier isValidDomain(string memory _tld) {
+    require(keccak256(bytes(_tld)) == keccak256(bytes("com")) || keccak256(bytes(_tld)) == keccak256(bytes("gov")), "Wrong domain level");
+    _;
+}
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol]: What is the purpose for this validator logic? -5

**Review by AndriianChestnykh — COMMENTED** [review]
Unfortunately, this implementation does not look good? -19 points for now

File: test/DomainRegistry.test.js (hunk 1 above)
Comment (nazarenkod) [reply test/DomainRegistry.test.js]: all comments removed

File: test/DomainRegistry.test.js (hunk 2 above)
Comment (nazarenkod) [reply test/DomainRegistry.test.js]: change to `domainName`

File: contracts/DomainRegistry.sol (hunk 3 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol]: removed tld from Domain struct

File: contracts/DomainRegistry.sol (hunk 4 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol]: removed `indexed` from `string` tld

File: contracts/DomainRegistry.sol (hunk 5 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:1]: - add DomainReleased and DomainCreated
- update Domain struct
- update tests

File: contracts/DomainRegistry.sol (hunk 6 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol]: - add validation for top level domains
- add tests

File: test/DomainRegistry.test.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+  beforeEach(async function () {
+    const hre = require("hardhat");
+    const DomainRegistry = await hre.ethers.getContractFactory("DomainRegistry");
+    domainRegistry = await DomainRegistry.deploy();
+    [owner] = await hre.ethers.getSigners();
+  });
+
+  it("Should register a new .com domain", async function () {
+    const domainForThisTest = "com";
+    const tx = await domainRegistry.connect(owner).registerDomain(domainForThisTest, {
```
Comment (AndriianChestnykh) [inline test/DomainRegistry.test.js:16]: Why do you `connect(owner)` each time? It is by default the first signer anyway and you just increase the code size.
-1 points

File: test/DomainRegistry.test.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        value: hre.ethers.parseEther("1.0"),
+    });
+    await tx.wait();
```
Comment (AndriianChestnykh) [inline test/DomainRegistry.test.js:19]: FYI Waiting for tx receipt is not mandatory for most cases in the tests

File: test/DomainRegistry.test.js
Comment (AndriianChestnykh) [inline test/DomainRegistry.test.js:1]: Why do you check your contract from only one user?
-2 points

File: contracts/DomainRegistry.sol [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    struct Domain {
+        uint256 deposit;
+        bool isRegistered;
+    }
+
+    mapping(string => Domain) public domains;
+
+    event DomainCreated(string tld, uint256 deposit);
+    event DomainReleased(string tld);
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:16]: What was you approach in choosing which fields to include in the events? Which of them to index or not?
How can it help with data monitoring?

File: contracts/DomainRegistry.sol [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+    modifier hasRequiredDeposit(uint256 _requiredDeposit) {
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:18]: Any reason to have an argument in the modifier? Can it work without it?

File: contracts/DomainRegistry.sol [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+contract DomainRegistry {
+    uint256 constant REQUIRED_DEPOSIT = 1 ether;
+
+    struct Domain {
+        uint256 deposit;
+        bool isRegistered;
+    }
+
+    mapping(string => Domain) public domains;
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:13]: Looking at this struct `Domain` and state var `domains` can't you see the problem? How can you answer which address in a domain controller?
-3 points

File: contracts/DomainRegistry.sol [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+    modifier domainDoesNotExist(string memory _tld) {
+        require(!domains[_tld].isRegistered, "Domain exists");
+        _;
+    }
+
+    modifier isTopLevelDomain(string memory _tld) {
+        require(bytes(_tld).length > 0, "Domain is empty");
+        require(bytes(_tld)[0] != bytes(".")[0], "Multilevel domains are not allowed");
+        require(bytes(_tld)[bytes(_tld).length - 1] != bytes(".")[0], "Multilevel domains are not allowed");
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:31]: What if point is in not at the first or last symbol in the string?

File: contracts/DomainRegistry.sol [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+    event DomainCreated(string tld, uint256 deposit);
+    event DomainReleased(string tld);
+
+    modifier hasRequiredDeposit(uint256 _requiredDeposit) {
+        require(msg.value >= _requiredDeposit, "Wrong eth amount");
+        _;
+    }
+
+    modifier domainDoesNotExist(string memory _tld) {
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:23]: It `_tld` the best arg name choice for someone who reads it for the first time?
-1 points

File: contracts/DomainRegistry.sol [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        emit DomainCreated(_tld, msg.value);
+    }
+
+    function releaseDomain(string memory _tld) public {
+        Domain storage domain = domains[_tld];
+        require(domain.isRegistered, "Domain isn't registered");    
+        uint256 depositAmount = domain.deposit;
+        domain.isRegistered = false;
+        domain.deposit = 0;
+        payable(msg.sender).transfer(depositAmount);
```
Comment (AndriianChestnykh) [inline contracts/DomainRegistry.sol:50]: So, anyone can release someone else's domain?
-3 points

File: contracts/DomainRegistry.sol (hunk 5 above)
Comment (AndriianChestnykh) [reply contracts/DomainRegistry.sol:1]: Unfortunately, all this does not help much the requirement
still -3 points

**Review by AndriianChestnykh — COMMENTED** [review]
This time it is somewhat better but far from good. Try better next time.
-12 points.

File: contracts/DomainRegistry.sol (hunk 14 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:50]: fxd

File: contracts/DomainRegistry.sol (hunk 13 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:23]: fxd

File: contracts/DomainRegistry.sol (hunk 12 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:31]: fxd

File: contracts/DomainRegistry.sol (hunk 11 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:13]: fxd

File: contracts/DomainRegistry.sol (hunk 10 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:18]: fxd

File: contracts/DomainRegistry.sol (hunk 9 above)
Comment (nazarenkod) [reply contracts/DomainRegistry.sol:16]: fxd

File: test/DomainRegistry.test.js
Comment (nazarenkod) [reply test/DomainRegistry.test.js:1]: add test and contract owner

File: test/DomainRegistry.test.js (hunk 8 above)
Comment (nazarenkod) [reply test/DomainRegistry.test.js:19]: fxd

**Pull request merged**